sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -569,10 +613,353 @@ public boolean getValidate() {
+          String jobIdToken,
+          String query,
+          String jsonQueryTempTable,
+          Boolean flattenResults,
+          BigQueryServices bqServices) {
+        return new BigQuerySource(
+            checkNotNull(jobIdToken, "jobIdToken"),
+            null, // jsonTable
+            checkNotNull(query, "query"),
+            checkNotNull(jsonQueryTempTable, "jsonQueryTempTable"),
+            checkNotNull(flattenResults, "flattenResults"),
+            checkNotNull(bqServices, "bqServices"));
+      }
+
+      static BigQuerySource fromTable(
+          String jobIdToken,
+          String jsonTable,
+          BigQueryServices bqServices) {
+        return new BigQuerySource(
+            checkNotNull(jobIdToken, "jobIdToken"),
+            checkNotNull(jsonTable, "jsonTable"),
+            null, // query
+            null, // jsonQueryTempTable
+            null, // flattenResults
+            checkNotNull(bqServices, "bqServices"));
+      }
+
+      private BigQuerySource(
+          String jobIdToken,
+          @Nullable String jsonTable,
+          @Nullable String query,
+          @Nullable String jsonQueryTempTable,
+          @Nullable Boolean flattenResults,
+          BigQueryServices bqServices) {
+        this.jobIdToken = checkNotNull(jobIdToken, "jobIdToken");
+        this.jsonTable = jsonTable;
+        this.query = query;
+        this.jsonQueryTempTable = jsonQueryTempTable;
+        this.flattenResults = flattenResults;
+        this.bqServices = checkNotNull(bqServices, "bqServices");
+      }
+
+      @Override
+      public List<BoundedSource<TableRow>> splitIntoBundles(
+          long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+        BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+        TableReference tableToExtract;
+        JobService jobService = bqServices.getJobService(bqOptions);
+        if (!Strings.isNullOrEmpty(query)) {
+          // Execute query
+          String queryJobId = jobIdToken + "-query";
+          tableToExtract = JSON_FACTORY.fromString(jsonQueryTempTable, TableReference.class);
+          executeQuery(queryJobId, query, tableToExtract, flattenResults, jobService);
+        } else {
+          tableToExtract = JSON_FACTORY.fromString(jsonTable, TableReference.class);
+        }
+        String extractJobId = jobIdToken + "-extract";
+        List<String> tempFiles = executeExtract(
+            extractJobId, tableToExtract, options.getTempLocation(), jobService);
+
+        TableSchema tableSchema = bqServices.getTableService(bqOptions).getTable(
+            tableToExtract.getProjectId(),
+            tableToExtract.getDatasetId(),
+            tableToExtract.getTableId()).getSchema();
+
+        return createSources(tempFiles, tableSchema);
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        // No way to estimate bytes, so returning 0.
+        return 0;
```

Comment:
can attempt to use the size in a query dry run or the size of the actual table. (Which can fail if validation is disabled, and then the runner will default to a sane value).